### DISCUSSION TITLE
xUnit tasks management

### DISCUSSION THREAD

**Discussion by pdufrene** [xunit/xunit#2797] [Question] [answered]
Hello all, 

I have a project with several tests inside using some tasks.
These unit tests are running well on my PC (Windows 10 Pro with a CPU with 14 cores and 20 logical cores).
But when I am executing the same Units test on Azure DevOps machine (I choose Windows2022 image), some tests are failing.
From what I was able to understand, it looks like some tasks are not executed and got delayed (until other tasks are completed maybe and in my case as I have some tasks producing data and other consuming these data, the timeout that I could have on BlockingCollection.Take for sample is failing)

I am guessing that the computer on Azure DevOps have lower CPU cores than my machine.
So any idea on how I can move on this issue ?
My first wish is to be able to reproduce the issue on my PC so I can investigate more.
Do you think there is a way to limit usage of number of Core on my PC. I tried maxThread = 1 or other values with out success.

**Comment by pdufrene** [comment]
To illustrate the issue, here a sample test:

```
public class ParallelTasksTest
{
    [Theory]
    [InlineData(19)]
    [InlineData(20)]
    public async Task ParallelTask(int taskCount)
    {
        var tasks = new List<Task>();
        var blockingCollections = new BlockingCollection<int>[taskCount];
        var values = new int[taskCount];

        for (var i = 0; i < taskCount; i++)
        {
            var blockingCollection = new BlockingCollection<int>();
            blockingCollections[i] = blockingCollection;
            var i1 = i;
            tasks.Add(Task.Run(() =>
            {
                if (!blockingCollection.TryTake(out var value, TimeSpan.FromSeconds(1)))
                {
                    throw new InvalidOperationException($"not in time: {i1}");
                }

                values[i1] = value;
            }));
        }

        for (var i = 0; i < taskCount; i++)
        {
            var i1 = i;
            tasks.Add(Task.Run(() =>
            {
                blockingCollections[i1].Add(i1);
            }));
        }

        await Task.WhenAll(tasks).ConfigureAwait(true);

        for (var i = 0; i < taskCount; i++)
        {
            blockingCollections[i].Dispose();
        }

        for (var i = 0; i < taskCount; i++)
        {
            Assert.Equal(i, values[i]);
        }
    }
}
```

On my Cpu which has 20 logical cores, the test with 19 tasks is passing while the one with 20 tasks is failing.
I agree that it's quite stupid to have 20 tasks inside a single unit tests.
But the CI machine has 1 or 2 cores (I don't know exactly) so this limit is quite annoying.

**Reply by bradwilson** [reply, marked as answer]
This does not appear to have anything to do with xUnit.net.

I did `dotnet new console` and copied your code into `Program.cs` with some minor modifications:

```csharp
using System.Collections.Concurrent;

var taskCount = Environment.ProcessorCount;

var tasks = new List<Task>();
var blockingCollections = new BlockingCollection<int>[taskCount];
var values = new int[taskCount];

for (var i = 0; i < taskCount; i++)
{
    var blockingCollection = new BlockingCollection<int>();
    blockingCollections[i] = blockingCollection;
    var i1 = i;
    tasks.Add(Task.Run(() =>
    {
        if (!blockingCollection.TryTake(out var value, TimeSpan.FromSeconds(1)))
        {
            throw new InvalidOperationException($"not in time: {i1}");
        }

        values[i1] = value;
    }));
}

for (var i = 0; i < taskCount; i++)
{
    var i1 = i;
    tasks.Add(Task.Run(() =>
    {
        blockingCollections[i1].Add(i1);
    }));
}

await Task.WhenAll(tasks).ConfigureAwait(true);

for (var i = 0; i < taskCount; i++)
{
    blockingCollections[i].Dispose();
}

for (var i = 0; i < taskCount; i++)
{
    if (i != values[i])
        throw new InvalidOperationException($"Expected {i}, got {values[i]}");
}
```

This code _sometimes_ fails with `dotnet run` just like the unit test does:

![image](https://github.com/xunit/xunit/assets/16944/afa04d87-e902-4379-a62b-5e62c07737a3)

The problem is `TimeSpan.FromSeconds(1)` is too small for the way that .NET grows the thread pool (and `Task.Run()` uses the default system thread pool by default). This write-up by @JonCole discusses it: https://gist.github.com/JonCole/e65411214030f0d823cb

In particular, this section:

> Once the number of existing (busy) threads hits the "minimum" number of threads, the ThreadPool will throttle the rate at which is injects new threads to one thread per 500 milliseconds. This means that if your system gets a burst of work needing an IOCP thread, it will process that work very quickly. However, if the burst of work is more than the configured "Minimum" setting, there will be some delay in processing some of the work as the ThreadPool waits for one of two things to happen 1. An existing thread becomes free to process the work 2. No existing thread becomes free for 500ms, so a new thread is created.

The default minimum worker threads (currently tested again .NET 7) is equal to the number of cores (returned by `Environment.ProcessorCount`).

So let's evaluate what happens when you have `n-1` vs. `n` taskCount (where `n` == your core count).

In the case of `n-1` (or 19 in your case), you ask for 19 background tasks to do work, and _they are all blocked waiting for a write that hasn't happened yet_. Then, you ask for 19 more background tasks to do the writes. We haven't hit 20 worker threads in-use yet, so what happens is when the 20th thread is launched (the first writer), it causes the first reader to free up and exit. This basically continues on down the line for the rest of the writer tasks, and each has at least one thread available to work with (in point of fact, a successful write frees up two threads, the reader and the writer, so the contention level goes down).

In the case of `n` (or 20 in your case), this time you ask for 20 background tasks to do work, and _they are all blocked waiting for a write that hasn't happened yet_. Now when you launch 20 more background tasks to do the writes, the system is going to pause for 500ms to see if any of the first 20 will free up. When they don't (they can't, by design of your test), it starts creating more worker threads, and then the writers start running, which then unblocks the log jam.

This is why it only _sometimes_ fails (for me, with 24 cores): you're waiting for 1000ms for the write, and you're guaranteed to have to wait _at least_ 500ms for the thread pool manager to decide to add more threads. I say at least for two reasons: (a) there is overhead to run the code, and especially the cost to spin up a new thread; (b) there is likely some imprecision in the waiting that the thread pool manager is doing, so that while it asks to wait 500ms, it will likely get the signal that 500ms has elapsed until sometime after 500ms has elapsed, and the amount of the difference in that time has a lot of factors that can't be counted on.

So you've basically perfectly created a scenario for "usually failing" by choosing a 1 second sleep up against a 500ms thread pool growth problem. If I change the sleep from 1 second to 250ms, then it's guaranteed to fail at `n` and now may sometimes fail at `n-1` simply because 250ms isn't all that much time. On the flip side, if I change the wait to 5 seconds, it's pretty much guaranteed to never fail, because the delta between 500ms and 5000ms is substantial.

This is an illustration of a race condition.

Now here's what you're really going to be up against: we're just testing with `n`. What happens when we have `5n` or `10n` instead? I can verify with `5n`, that even sleeping 5 seconds is not sufficient for me with an otherwise not busy R9 5900X. This is because the thread pool grows slowly (and if you're smart, you won't try to predict how slowly it grows, or by how much, because you're just going to be crossing your fingers that you've chosen enough sleep time). Your CI machines with 1 or 2 cores are going to be waiting a theoretically looooooooooong time for the growth of the thread pool.

In my opinion, you either need to restructure your unit tests so that you're not subject to the delays, or you're going to end up needing to implement your own thread pool which is guaranteed to be the correct size, and running the background work on that thread pool. (There are probably other potential solutions, but those are the first two that popped into my head.)

**Reply by pdufrene** [reply]
@bradwilson Ok thank you, it makes perfectly sense. Very clear and detailed explanation by the way! Thank you for taking time to do it.

**Reply by bradwilson** [reply]
Another option you might consider, if it's available to you, is to use operations which do not block the thread. What you'd be looking for is async (that is, `Task` or `ValueTask` returning) operations.  Unfortunately, `BlockingCollection` does not have any such operations on it (I realize that `BlockingCollection` is standing in for some other operation, that may or may not have async variants).

There is another way you could try to accomplish this, without something that's already built in to be async.
Instead of just doing a single `TryTake` with your whole timeout (1000ms), you could do it in small slices, and interleave calls to `Task.Yield` to return your thread so anything else that's waiting can complete. Here's a version of your reader loop that splits the total wait time into 10ms slices:

```csharp
var sliceSize = 10;
var sliceTimeSpan = TimeSpan.FromMilliseconds(sliceSize);

tasks.Add(Task.Run(async () =>
{
    var waitTimeLeft = 1000;

    while (waitTimeLeft > 0)
    {
        if (blockingCollection.TryTake(out var value, sliceTimeSpan))
        {
            values[i1] = value;
            return;
        }

        waitTimeLeft -= sliceSize;
        await Task.Yield();
    }

    throw new InvalidOperationException($"not in time: {i1}");
}));
```

With this change, I was able to successfully run even as many as `Environment.ProcessorCount * 100` tasks. You can tweak the size of the slice as appropriate to balance performance vs. CPU overhead (the 10ms overhead wait time feels small enough so that we're not wasting too much "do nothing" time waiting, but large enough so that we're not just wasting time looping over and over again).

Running with `n` tasks, it finishes for me in about 1.25 seconds; running with `100n` tasks, it finishes for me in about 2.5 seconds. That's not too bad in my book. I'm guessing that the thread pool size either does not grow, or grows so little that it makes no difference, once we hit the minimum threshold. This is because each thread is only ever "busy" for 10ms before yielding itself back, so it's unlikely the system ever has to wait for 500ms for a worker thread to become available. The difference in 1.25 seconds vs. 2.5 seconds is almost assuredly all down to the overhead of creating 100x more Tasks to be scheduled amongst the worker threads.
